feat: Process transpile path

amoshydra opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
<!--- If it resolves an open issue, please link to the issue here. For example "Resolves: #1337" -->
This PR aimed to resolve issue #4334 by processing the plain string entry of `options.build.transpile`.

The processing includes:
1. convert given path to os friendly file path (using path.normalize)
2. sanitize string for RegExp (using lodash escapeRegExp)
3. convert the sanitized string to RegExp (original)
4. use this RegExp for testing (original)

However, doing this conflict with one of the test case that uses escaped RegExp string as entry. 
**input**: `'vue\\.test\\.js'`
**test**: `node_modules/vue.test.js`
This will no longer work (thus breaking change) because we will now be expecting user to write plain string instead.

Unfortunally this test is also depended on the OS as it uses `path.normalize` which its behaviour is depended on OS.

## Checklist:
<!--- Put an `x` in all the boxes that apply. -->
<!--- If your change requires a documentation PR, please link it appropriately -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

amoshydra committed:
Convert and sanitize transpile path

amoshydra committed:
Normalize path to os frienly path before testing

amoshydra committed:
Add test for scoped module

amoshydra committed:
Remove the need to espace plain string for regex

amoshydra committed:
Replace path join with path normalize

clarkdo commented on test/unit/basic.dev.test.js:
```diff
@@ -60,10 +62,11 @@ describe('basic dev', () => {
+    expect(transpile(path.normalize('node_modules/vue-test'))).toBe(true)
+    expect(transpile(path.normalize('node_modules/vue.test.js'))).toBe(true)
+    expect(transpile(path.normalize('node_modules/test.vue.js'))).toBe(true)
+    expect(transpile(path.normalize('node_modules/@scoped/packageA/src/index.js'))).toBe(true)
```
We should also mock windows path, like 'node_modules\\\\@scoped\\\\packageA\\\\src\\\\index.js'

amoshydra replied:
Hi @clarkdo, I am not sure how I can mock windows path here and still have the test run correctly on both POSIX and Windows due to the single direction POSIX to Windows normalization by `path.normalize`.

Also, in all case (I suspect?) babel will always pass native filepath into the exclude function automatically.

amoshydra replied:
> Also, in all case (I suspect?) babel will always pass native filepath into the exclude function automatically.

I meant to imply that: Since native filepath is used, we should be able to cover both case for 
Windows and POSIX using (ci.azure) and (ci/circleci: test-unit).

Is that correct?

clarkdo replied:
Actually not, exclude file path will be `c:\\project\\node_modules\\@Scoped\\packageA\\src\\index.js`.

So we may also need to normalize `file` in `exclude`.

That's also why I mentioned we'd better to mock windows path, like:

```js
transpile('node_modules\\\\@scope\\\\module')
```

clarkdo replied:
If we don't consider compatibility, we can only normalize `file` in `exclude`.

amoshydra replied:
Hi @clarkdo, sorry I may not have understood on what you meant by compatibility, I am still unclear about why we need to normalize the filepath (`file`) in the `exclude` function.

## Clarifications:
1. Just to be clear, when we talk about `normalize`, this is what we mean right?
    > to convert a given file path to a new plain string with the OS's separator
2. I do understand that the filepath in the exclude function will be `"c:\\project\\node_modules\\@scoped\\packageA\\src\\index.js"` in Windows.
However, I am not sure about why we should do `transpile('node_modules\\\\@scope\\\\module')` in the test.
    1. Firstly, this will always fail the test for POSIX system, since our entry in options.build.transpile will be normalized to test for `/`.

    2. This will always fail for Windows too since we are testing `/@scope\\module/` against `"@scope\\\\module"`
    because:
    `/@scope\\module/.test("@scope\\\\module") === false`

    3. Using 4 backslash seems to imply that you want to convert the string to a RegExp? i.e. 
    `new RegExp('@scope\\\\module').source === /@scope\\module/.source`

    Therefore, the more accurate behavior will be to do `transpile(path.normalize('node_modules/@scope/module'))` instead? as this will create the same behavior as how webpack passes filepath to the `exclude` function.




## My thoughts:
https://github.com/nuxt/nuxt.js/blob/3702dfe5dcb5c4fdeb03fcd3025f5a5ae39232be/packages/webpack/src/config/base.js#L173-L176

According to my understanding, `exclude` function will receive a `filepath` as parameter. The file path will be constructed depending on the OS. On Windows it will always be separated with `\` on POSIX it will be seperated with `/`.

`modulesToTranspile` will contain RegExps using the file path we specified in the `options.build.transpile` array. These RegExps' file path will be normalize according to the OS.

Since e85a201, regardless of the format that user specify (POSIX or Windows) they will always get normalized to the OS's format. This will match well with the logic in exclude function.

## Illustration of logic:
#### Windows

| input                | moduleToTranspile    | exclude's param |
| :--                  | :--                  | :--             |
| `"@scoped/packageA"`  | `/@scoped\\packageA/` | `"c:\\project\\node_modules\\@scoped\\packageA\\src\\index.js"`
| `"@scoped\\packageA"` | `/@scoped\\packageA/` | `"c:\\project\\node_modules\\@scoped\\packageA\\src\\index.js"`

#### POSIX

| input                | moduleToTranspile    | exclude's param |
| :--                  | :--                  | :--             |
| `"@scoped/packageA"`  | `/@scoped\/package/` | `"/project/node_modules/@scoped/packageA/src/index.js"`
| `"@scoped\\packageA"` | `/@scoped\/package/` | `"/project/node_modules/@scoped/packageA/src/index.js"`

## Tested
Both of these cases are written as part of the test case too
https://github.com/nuxt/nuxt.js/blob/e85a201be01ca8dcd8e59f4fcc326cf2b501b323/test/unit/basic.dev.test.js#L29-L34

and both passed the unit test ran in Windows (in [ci.azure](https://dev.azure.com/nuxt/nuxt.js/_build/results?buildId=612&view=logs)) and POSIX environment (in [circleci](https://circleci.com/gh/nuxt/nuxt.js/15562?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link))

---

Thank for you reading through my long comment. :)

clarkdo replied:
Thanks you the clarifications, it works fine after converting POSIX commit.

The `\\\\` is because github will escape the `\` in quote, so I add duplicate ones in whole comment, sorry.

clarkdo commented on packages/config/src/options.js:
```diff
@@ -320,7 +321,12 @@ export function getNuxtConfig(_options) {
+    .map(module => (module instanceof RegExp)
+      ? module
+      : new RegExp(
+        escapeRegExp(path.normalize(module))
```
`normalize` won't parse windows style path into POSIX, so it may still not work in windows.

amoshydra replied:
@clarkdo,

I think you meant to say "it may still not work in POSIX"?

Here is my proposal for this issue: e85a201
>  Convert to POSIX path before normalisation
>  
>  path.normalize can convert POSIX to Windows but not Windows to POSIX.
>  This replacement is added to ensure that the path is always in POSIX
>  form before normalisation occur.

clarkdo replied:
Please see above comment 😄

amoshydra committed:
Add test for backslashed path

amoshydra committed:
Convert to POSIX path before normalisation

path.normalize can convert POSIX to Windows but not Windows to POSIX.
This replacement is added to ensure that the path is always in POSIX
form before normalisation occur.
